v2/spanner-to-sourcedb/src/main/java/com/google/cloud/teleport/v2/templates/dbutils/connection/CassandraConnectionHelper.java

```diff
@@ -0,0 +1,192 @@
+ *
+ * <p>Typical usage:
+ *
+ * <pre>
+ *   CassandraConnectionHelper helper = new CassandraConnectionHelper();
+ *   helper.init(connectionHelperRequest);
+ *   CqlSession session = helper.getConnection(connectionKey);
+ * </pre>
+ */
+public class CassandraConnectionHelper implements IConnectionHelper<CqlSession> {
+
+  /** Logger for logging information and errors. */
+  private static final Logger LOG = LoggerFactory.getLogger(CassandraConnectionHelper.class);
+
+  /** A thread-safe connection pool storing {@link CqlSession} instances mapped to unique keys. */
+  private static Map<String, CqlSession> connectionPoolMap = new ConcurrentHashMap<>();
+
+  /**
+   * Initializes the connection pool with connections for the provided Cassandra shards.
+   *
+   * @param connectionHelperRequest The request object containing shard details and connection
+   *     settings.
+   * @throws IllegalArgumentException if any shard validation fails or invalid shard types are
+   *     provided.
+   */
+  @Override
+  public synchronized void init(ConnectionHelperRequest connectionHelperRequest) {
+    if (connectionPoolMap != null && !connectionPoolMap.isEmpty()) {
+      LOG.info("Connection pool is already initialized.");
+      return;
+    }
+
+    LOG.info(
+        "Initializing Cassandra connection pool with size: {}",
+        connectionHelperRequest.getMaxConnections());
+
+    List<Shard> shards = connectionHelperRequest.getShards();
+    for (Shard shard : shards) {
+      if (!(shard instanceof CassandraShard)) {
+        LOG.error("Invalid shard type: {}", shard.getClass().getSimpleName());
+        throw new IllegalArgumentException("Invalid shard object");
+      }
+
+      CassandraShard cassandraShard = (CassandraShard) shard;
+      try {
+        cassandraShard.validate();
+        CqlSession session = createCqlSession(cassandraShard);
+        String connectionKey = generateConnectionKey(cassandraShard);
+        connectionPoolMap.put(connectionKey, session);
+        LOG.info("Connection initialized for key: {}", connectionKey);
+      } catch (Exception e) {
+        LOG.error("Failed to initialize connection for shard: {}", cassandraShard, e);
+      }
+    }
+  }
+
+  /**
+   * Retrieves a {@link CqlSession} connection from the connection pool.
+   *
+   * @param connectionRequestKey The unique key identifying the connection in the pool.
+   * @return The {@link CqlSession} instance associated with the given key.
+   * @throws ConnectionException If the connection pool is not initialized or no connection is found
+   *     for the key.
+   */
+  @Override
+  public CqlSession getConnection(String connectionRequestKey) throws ConnectionException {
+    if (connectionPoolMap == null || connectionPoolMap.isEmpty()) {
+      LOG.warn("Connection pool not initialized.");
+      throw new ConnectionException("Connection pool is not initialized.");
+    }
+
+    CqlSession session = connectionPoolMap.get(connectionRequestKey);
```

Comment:
This implementation seems to be keeping the session cached in static memory for the duration of the of the process.
We might want to consider adding checks to make sure the session is still live.

In case you add those checks and you need to refresh a stale session, you might need to use a threadsafe hash function, you could make also make getConnection synchronous instead as anyways all dataflow works will be contending for the same session object here.